metric/system/cgroup/util.go

```diff
@@ -275,6 +277,22 @@ the container as /sys/fs/cgroup/unified and start the system module with the hos
 				controllerPath = r.rootfsMountpoint.ResolveHostFS(filepath.Join("/sys/fs/cgroup/unified", path))
 			}
 
+			// Check if there is an entry for controllerPath already cached.
+			if tmp, ok := r.v2ControllerPathCache.Load(controllerPath); ok {
+				cacheEntry, ok := tmp.(pathListWithTime)
+				if ok {
+					// If the cached entry for controllerPath is not older than 5 minutes,
+					// return the cached entry.
+					if time.Since(cacheEntry.added) < time.Duration(5*time.Minute) {
+						cPaths.V2 = cacheEntry.pathList.V2
+						continue
+					}
+				}
+				// Consider the existing entry for controllerPath invalid. The entry can
+				// (1) not be casted to pathListWithTime or (2) is older than 5 minutes.
+				r.v2ControllerPathCache.Delete(controllerPath)
```

Comment:
Instead of using a `sync.Map` why not use a struct wrapping regular map and a mutex? This would give us better type safety and we wouldn't need the type assertion in this code here. It should also make the logic simpler here, as it would be just about whether the cache entry has expired or not.

Reply:
I did introduce the synchronization mechanism  as it is not guaranteed that `Reader` is only used in a sequential order. And as multiple Go routine can use `Reader` and a regular `map` is not Go thread safe, I went with `sync.Map`. I did choose `sync.Map` over a struct with a regular map and a mutex, as it is harder to abuse and therefore more safe. While I agree on the added complexity around `sync.Map` handling, it prevents direct accessing a map in a struct, without holding the sync mutex.
Is this ok for you?

Reply:
Totally get the need for a synchronization here!  It's just that the documentation of `sync.Map` itself doesn't recommend using it over a regular map with a mutex, citing type safety as one of the reasons :)

https://pkg.go.dev/sync#Map

> The Map type is specialized. Most code should use a plain Go map instead, with separate locking or coordination, for better type safety and to make it easier to maintain other invariants along with the map content.

I also searched through the existing `elastic-agent-system-metrics` and `elastic-agent` codebases and I'm not seeing any uses of `sync.Map`, probably for this same reason.  So I'm hesitant to break that pattern here.

Reply:
While `sync.Map` is not used in the two named repositories, other beats and Elastic projects are using it - https://github.com/search?type=code&q=org%3Aelastic+lang%3Ago+sync.Map

I updated the PR to keep the consistency of `elastic-agent-system-metrics`